lib/Dialect/FIRRTL/Transforms/LowerIntmodules.cpp

```diff
@@ -0,0 +1,188 @@
+        Value result;
+        BundleType::BundleElement element;
+      };
+      SmallVector<OutputInfo> outputs;
+      for (auto [idx, result] : llvm::enumerate(inst.getResults())) {
+        // Replace inputs with wires that will be used as operands.
+        if (inst.getPortDirection(idx) != Direction::Out) {
+          auto w = builder.create<WireOp>(result.getLoc(), result.getType())
+                       .getResult();
+          result.replaceAllUsesWith(w);
+          inputs.push_back(w);
+          continue;
+        }
+
+        // Gather outputs.  This will become a bundle if more than one, but
+        // typically there are zero or one.
+        auto ftype = dyn_cast<FIRRTLBaseType>(inst.getType(idx));
+        if (!ftype) {
+          inst.emitError("intrinsic has non-FIRRTL or non-base port type")
+              << inst.getType(idx);
+          signalPassFailure();
+          return;
+        }
+        outputs.push_back(
+            OutputInfo{inst.getResult(idx),
+                       BundleType::BundleElement(inst.getPortName(idx),
+                                                 /*isFlip=*/false, ftype)});
+      }
+
+      // Create the replacement operation.
+      if (outputs.empty()) {
+        // If no outputs, just create the operation.
+        builder.create<GenericIntrinsicOp>(/*result=*/Type(),
+                                           op.getIntrinsicAttr(), inputs,
+                                           op.getParameters());
+
+      } else if (outputs.size() == 1) {
+        // For single output, the result is the output.
+        auto resultType = outputs.front().element.type;
+        auto intop = builder.create<GenericIntrinsicOp>(
+            resultType, op.getIntrinsicAttr(), inputs, op.getParameters());
+        outputs.front().result.replaceAllUsesWith(intop.getResult());
+        // auto name = builder.getStringAttr(inst.getInstanceName() + "_" +
+        // outputs.front().element.name.strref()); intop->setAttr("name", name);
+      } else {
+        // For multiple outputs, create a bundle with fields for each output
+        // and replace users with subfields.
+        auto resultType = builder.getType<BundleType>(llvm::map_to_vector(
+            outputs, [](const auto &info) { return info.element; }));
+        auto intop = builder.create<GenericIntrinsicOp>(
+            resultType, op.getIntrinsicAttr(), inputs, op.getParameters());
+        for (auto &output : outputs)
+          output.result.replaceAllUsesWith(builder.create<SubfieldOp>(
+              intop.getResult(), output.element.name));
+        // intop->setAttr("name", inst.getInstanceNameAttr());
+      }
+      // Remove instance from IR and instance graph.
+      use->erase();
+      inst.erase();
+    }
+    // Remove intmodule from IR and instance graph.
+    ig.erase(node);
+    op.erase();
+  }
+
+  // Special handling for magic EICG wrapper extmodule.  Deprecate and remove.
+  if (fixupEICGWrapper) {
```

Comment:
nit: It's unfortunate that we have to duplicate similar code just for single use case. Do you happen to know it's already safe to remove the flag? Or should we just define the EICG wrapper as an intrinsic?

Reply:
We still haven't flipped the switch on this. @mwachs5 has been working on a final missing piece in `ExtractInstances`, where we're not yet properly handling the intrinsic. Once that's in, we can enable this option by default for a few weeks, and then declare it canon and rip out the old `EICG_wrapper` handling everywhere. But I guess this will stay for a few more months until all Chisel code bases are migrated over to using clock gating intrinsics.

Reply:
That makes sense. Thank you for elaborating this!

Reply:
Awesome thanks for the summary and explanation!

FWIW I intentionally didn't maintain this as some sort of generic extmodule support (or try to combine for code re-use, as not-awesome as that is) specifically because I want this to be super easy to delete when the time comes (and not easily become more load-bearing, standing out as a weird thing we wish we didn't do) 👍 .

Special weird magic handling does not sully the heart of the pass and must sit in the corner with a dunce cap on it 😉 .  Anyway should be able to drop this "soon"!